graph/path/morphism_apply_functions.go

```diff
@@ -20,58 +20,80 @@ import (
 	"github.com/google/cayley/quad"
 )
 
+// join puts two iterators together by intersecting their result sets with an AND
+// Since we're using an and iterator, it's a good idea to put the smallest result
+// set first so that Next() produces fewer values to check Contains()
+func join(qs graph.QuadStore, itL, itR graph.Iterator) graph.Iterator {
+	and := iterator.NewAnd(qs)
+	and.AddSubIterator(itL)
+	and.AddSubIterator(itR)
+
+	return and
+}
+
+// isMorphism represents all nodes passed in-- if there are none, this function
+// acts as a passthrough for the previous iterator
 func isMorphism(nodes ...string) morphism {
 	return morphism{
 		Name:     "is",
 		Reversal: func() morphism { return isMorphism(nodes...) },
-		Apply: func(qs graph.QuadStore, it graph.Iterator) graph.Iterator {
-			var sub graph.Iterator
+		Apply: func(qs graph.QuadStore, in graph.Iterator) graph.Iterator {
 			if len(nodes) == 0 {
-				sub = qs.NodesAllIterator()
-			} else {
-				fixed := qs.FixedIterator()
-				for _, n := range nodes {
-					fixed.Add(qs.ValueOf(n))
-				}
-				sub = fixed
+				// Acting as a passthrough here is equivalent to
+				// building a NodesAllIterator to Next() or Contains()
+				// from here as in previous versions
+				return in
 			}
-			and := iterator.NewAnd(qs)
-			and.AddSubIterator(sub)
-			and.AddSubIterator(it)
-			return and
+
+			isNodes := qs.FixedIterator()
+			for _, n := range nodes {
+				isNodes.Add(qs.ValueOf(n))
+			}
+
+			// Anything with fixedIterators will usually have a much
+			// smaller result set, so join isNodes first here
+			return join(qs, isNodes, in)
 		},
 	}
 }
 
+// hasMorphism is
```

Comment:
Fixed. In my local copy this function got renamed to reaches, but I didn't want to break the API via renaming exported functions here. This one in particular is tricky to describe as Has-- thanks for the current review, and please let me know if I've misunderstood the logic this represents.